### PR TITLE
Utilize exact kNN search when gathering k > numVectors in a segment

### PR DESCRIPTION
**Pull request by benwtrent**
When requesting for k >= numVectors, it doesn't make sense to go through the HNSW graph. Even without a user supplied filter, we should not explore the HNSW graph if it contains fewer than `k` vectors.

One scenario where we may still explore the graph if k >= numVectors is when not every document has a vector and there are deleted docs. But, this commit significantly improves things regardless.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java [hunk 1]
```diff
@@ -171,13 +177,13 @@ protected TopDocs exactSearch(LeafReaderContext context, DocIdSetIterator accept
     }
 
     VectorScorer vectorScorer = createVectorScorer(context, fi);
+    // Iterate over the vector docs that match the filter and find the top k
+    DocIdSetIterator iterator =
+        ConjunctionUtils.intersectIterators(List.of(acceptIterator, vectorScorer.iterator()));
```
Comment (benwtrent) [inline lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java:182]: An intersection of the vector values and the user provided filter seemed like the way it should have always been. 

Do we want to remove the `fieldExists` query logic when a user provides a pre-filter?

File: lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java [hunk 2]
```diff
@@ -110,6 +110,12 @@ private TopDocs getLeafResults(LeafReaderContext ctx, Weight filterWeight) throw
     int maxDoc = ctx.reader().maxDoc();
 
     if (filterWeight == null) {
+      int cost = liveDocs == null ? maxDoc : liveDocs.length();
```
[the remark this answers is not in the document]
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java:113]: liveDocs.length() is always equal to `maxDoc`

**Review by jpountz — COMMENTED** [review]
The idea makes sense to me, what is less clear to me is whether this logic belongs to the Query or to the vector reader: should `searchNearestNeighbors` implicitly do a linear scan when `k` is greater than `size()`?

**Comment by benwtrent** [comment]
> The idea makes sense to me, what is less clear to me is whether this logic belongs to the Query or to the vector reader: should searchNearestNeighbors implicitly do a linear scan when k is greater than size()?

I like that idea, we should do that.

File: lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsReader.java [hunk 3]
```diff
@@ -238,11 +238,23 @@ public void search(String field, float[] target, KnnCollector knnCollector, Bits
...
-        new OrdinalTranslatedKnnCollector(knnCollector, scorer::ordToDoc),
-        getGraph(fieldEntry),
-        scorer.getAcceptOrds(acceptDocs));
+    if (knnCollector.k() < scorer.maxOrd()) {
+      HnswGraphSearcher.search(
+          scorer,
+          new OrdinalTranslatedKnnCollector(knnCollector, scorer::ordToDoc),
+          getGraph(fieldEntry),
+          scorer.getAcceptOrds(acceptDocs));
+      return;
```
Comment (jpountz) [inline lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsReader.java:247]: nit: from a style perspective, I'd rather not return here and have an `else` block instead

**Review by jpountz — APPROVED** [review]
Thanks, I find that it also looks nicer, and it saves one `IndexInput.clone()` as well.

**Review by jpountz — APPROVED** [review]

**Pull request merged**